ledger/src/shred/merkle.rs

```diff
@@ -687,21 +706,29 @@ fn make_merkle_proof(
     mut index: usize, // leaf index ~ shred's erasure shard index.
     mut size: usize,  // number of leaves ~ erasure batch size.
     tree: &[Hash],
-) -> Option<Vec<&MerkleProofEntry>> {
-    if index >= size {
-        return None;
-    }
+) -> impl Iterator<Item = Result<&MerkleProofEntry, Error>> {
     let mut offset = 0;
-    let mut proof = Vec::<&MerkleProofEntry>::new();
-    while size > 1 {
-        let node = tree.get(offset + (index ^ 1).min(size - 1))?;
-        let entry = &node.as_ref()[..SIZE_OF_MERKLE_PROOF_ENTRY];
-        proof.push(<&MerkleProofEntry>::try_from(entry).unwrap());
-        offset += size;
-        size = (size + 1) >> 1;
-        index >>= 1;
+    if index >= size {
+        // Force below iterator to return Error.
+        (size, offset) = (0, tree.len());
     }
-    (offset + 1 == tree.len()).then_some(proof)
+    std::iter::from_fn(move || {
+        if size > 1 {
+            let Some(node) = tree.get(offset + (index ^ 1).min(size - 1)) else {
+                return Some(Err(Error::InvalidMerkleProof));
+            };
+            offset += size;
+            size = (size + 1) >> 1;
+            index >>= 1;
+            let entry = &node.as_ref()[..SIZE_OF_MERKLE_PROOF_ENTRY];
+            let entry = <&MerkleProofEntry>::try_from(entry).unwrap();
+            Some(Ok(entry))
+        } else if offset + 1 == tree.len() {
```

Comment:
Looks correct, based on `get_merkle_tree_size` when `size` reaches `1`, the `offset` will always fulfill this condition.